registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp

```diff
@@ -116,90 +116,57 @@ estimateRigidTransformation (const pcl::PointCloud<PointSource> &cloud_src,
 //////////////////////////////////////////////////////////////////////////////////////////////
 template <typename PointSource, typename PointTarget, typename Scalar> inline void
 pcl::registration::TransformationEstimationSymmetricPointToPlaneLLS<PointSource, PointTarget, Scalar>::
-constructTransformationMatrix (const Vector6d &parameters,
+constructTransformationMatrix (const Vector6 &parameters,
                                Matrix4 &transformation_matrix) const
 {
   // Construct the transformation matrix from rotation and translation 
-  Eigen::Matrix<Scalar, 3, 3> rotation_matrix;
-  rotation_matrix = Eigen::AngleAxis<Scalar> (parameters (2), Eigen::Matrix<Scalar, 3, 1>::UnitZ ()) * 
-                    Eigen::AngleAxis<Scalar> (parameters (1), Eigen::Matrix<Scalar, 3, 1>::UnitY ()) *
-                    Eigen::AngleAxis<Scalar> (parameters (0), Eigen::Matrix<Scalar, 3, 1>::UnitX ());
-  Eigen::Matrix<Scalar, 3, 1> translation_vector;
-  translation_vector << static_cast<Scalar> (parameters (3)), static_cast<Scalar> (parameters (4)), static_cast<Scalar> (parameters (5));
-  transformation_matrix.setIdentity();
-  transformation_matrix.template block<3, 3> (0, 0) = rotation_matrix;
-  transformation_matrix.template block<3, 1> (0, 3) = translation_vector;
+  const Eigen::AngleAxis<Scalar> rotation_z (parameters (2), Eigen::Matrix<Scalar, 3, 1>::UnitZ ());
+  const Eigen::AngleAxis<Scalar> rotation_y (parameters (1), Eigen::Matrix<Scalar, 3, 1>::UnitY ());
+  const Eigen::AngleAxis<Scalar> rotation_x (parameters (0), Eigen::Matrix<Scalar, 3, 1>::UnitX ());
+  const Eigen::Translation<Scalar, 3> translation (parameters (3), parameters (4), parameters (5));
+  const Eigen::Transform<Scalar, 3, Eigen::Affine> transform = rotation_z * rotation_y * rotation_x * translation * rotation_z * rotation_y * rotation_x;
+  transformation_matrix = transform.matrix();
 }
 
 //////////////////////////////////////////////////////////////////////////////////////////////
 template <typename PointSource, typename PointTarget, typename Scalar> inline void
 pcl::registration::TransformationEstimationSymmetricPointToPlaneLLS<PointSource, PointTarget, Scalar>::
 estimateRigidTransformation (ConstCloudIterator<PointSource>& source_it, ConstCloudIterator<PointTarget>& target_it, Matrix4 &transformation_matrix) const
 {
-  using Matrix6d = Eigen::Matrix<double, 6, 6>;
+  using Matrix6 = Eigen::Matrix<Scalar, 6, 6>;
+  using Vector3 = Eigen::Matrix<Scalar, 3, 1>;
 
-  Matrix6d ATA;
-  Vector6d ATb;
+  Matrix6 ATA;
+  Vector6 ATb;
   ATA.setZero ();
   ATb.setZero ();
-  auto U = ATA.triangularView<Eigen::Upper> ();
+  auto M = ATA.template selfadjointView<Eigen::Upper> ();
 
   // Approximate as a linear least squares problem
   for (; source_it.isValid () && target_it.isValid (); ++source_it, ++target_it)
   {
-    if (!isFinite(*source_it) ||
-        !isFinite(*target_it))
+    Vector3 s (source_it->x, source_it->y, source_it->z);
```

Comment:
I notice that the values are not mean subtracted. Will that impact convergence or numerical stability as hinted in the paper? (Eq 10 is EM for mean-subtracted set of points, rest seems similar)

Reply:
Fair point. The question is, who and when should take care of demeaning. I don't see an efficient way to perform it within this class. And generally, mean subtraction is a good idea for every objective, not only symmetric. So I feel that there should be a general mechanism to do this though it's outside of the scope of this PR then.

Reply:
There's no need to perform demeaning. Find the mean in a separate pass (or get it as parameter, defaulted to `Eigen3X::Zero()` and `const auto s = Vector3(...) - mean;` No need to add the mean

Reply:
I've subtracted the mean from the point clouds in order to improve numerical stability/convergence.